applications/FemToDemApplication/custom_elements/femdem2d_element.cpp

```diff
@@ -128,203 +128,96 @@ void FemDem2DElement::FinalizeSolutionStep(ProcessInfo &rCurrentProcessInfo)
-	noalias(J[0]) = ZeroMatrix(dimension, dimension);
-	J = GetGeometry().Jacobian(J, mThisIntegrationMethod, delta_position);
+	//resizing as needed the LHS
+	unsigned int mat_size = number_of_nodes * (dimension) ; // water pressure
 
-	WeakPointerVector<Element> &elem_neigb = this->GetValue(NEIGHBOUR_ELEMENTS);
-	KRATOS_ERROR_IF(elem_neigb.size() == 0) << " Neighbour Elements not calculated" << std::endl;
+	if ( rLeftHandSideMatrix.size1() != mat_size )
+	rLeftHandSideMatrix.resize( mat_size, mat_size, false );
 
-	// Loop Over Integration Points
-	for (unsigned int PointNumber = 0; PointNumber < integration_points.size(); PointNumber++) {
-		Matrix InvJ(dimension, dimension);
-		noalias(InvJ) = ZeroMatrix(dimension, dimension);
-		double detJ = 0;
-		MathUtils<double>::InvertMatrix(J[PointNumber], InvJ, detJ);
+	noalias(rLeftHandSideMatrix) = ZeroMatrix(mat_size, mat_size); //resetting LHS
+	if (rRightHandSideVector.size() != mat_size )
+	rRightHandSideVector.resize(mat_size, false);
 
-		double integration_weight = integration_points[PointNumber].Weight() * detJ;
-		integration_weight *= this->GetProperties()[THICKNESS];
+	noalias(rRightHandSideVector) = ZeroVector(mat_size); //resetting RHS
+	
+	//create and initialize element variables:
+	ElementDataType Variables;
+	this->InitializeElementData(Variables, rCurrentProcessInfo);
 
-		KRATOS_ERROR_IF(detJ < 0) << " SMALL DISPLACEMENT ELEMENT INVERTED: |J|<0 " << std::endl;
+	//create constitutive law parameters:
+	ConstitutiveLaw::Parameters Values(this->GetGeometry(), this->GetProperties(), rCurrentProcessInfo);
 
-		//compute cartesian derivatives for this integration point  [dN/dx_n]
-		noalias(DN_DX) = prod(DN_De[PointNumber], InvJ);
+	//set constitutive law flags:
+	Flags& ConstitutiveLawOptions=Values.GetOptions();
 
-		//set shape functions for this integration point
-		Vector N = row(Ncontainer, PointNumber);
+	ConstitutiveLawOptions.Set(ConstitutiveLaw::COMPUTE_STRESS);
+	ConstitutiveLawOptions.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR);
 
-		//b.-compute infinitessimal strain
-		this->CalculateInfinitesimalStrain(strain_vector, DN_DX);
+	//reading integration points
+	const GeometryType::IntegrationPointsArrayType& integration_points = GetGeometry().IntegrationPoints(mThisIntegrationMethod );
 
-		ConstitutiveLaw::Parameters Values(GetGeometry(), GetProperties(), rCurrentProcessInfo);
+	//auxiliary terms
+	Vector VolumeForce(dimension);
+	noalias(VolumeForce) = ZeroVector(dimension);
 
-		//set constitutive law variables: (it passes only references to this local variables)
-		Values.SetStrainVector(strain_vector);
-		Values.SetStressVector(stress_vector);
-		Values.SetConstitutiveMatrix(constitutive_matrix);
-		Values.SetShapeFunctionsDerivatives(DN_DX);
-		Values.SetShapeFunctionsValues(N);
-		//values to be set:
-		Values.SetDeterminantF(detF);
-		Values.SetDeformationGradientF(F);
+	WeakPointerVector<Element> &elem_neigb = this->GetValue(NEIGHBOUR_ELEMENTS);

```

Comment:
I recommend to use auto, @RiccardoRossi is refactoring the neighbors stuff, will avoid problems 
```suggestion
	auto& elem_neigb = this->GetValue(NEIGHBOUR_ELEMENTS);
```